landlab/components/flow_accum/lossy_flow_accumulator.py

```diff
@@ -0,0 +1,464 @@
+                    raise TypeError(
+                        "The loss_function should take a float, and return "
+                        "a float."
+                    )
+                # now, for logical consistency in our calls to
+                # find_drainage_area_and_discharge, wrap the func so it has two
+                # arguments:
+
+                def lossfunc(Qw, dummyn, dummyl, dummygrid):
+                    return float(loss_function(Qw))
+
+                self._lossfunc = lossfunc
+
+            elif num_params == 2:
+                # check the func takes a single value and turns it into a new
+                # single value:
+                if not isinstance(loss_function(1., 0), float):
+                    raise TypeError(
+                        "The loss_function should take (float, int), and "
+                        "return a float."
+                    )
+                # now, for logical consistency in our calls to
+                # find_drainage_area_and_discharge, wrap the func so it has two
+                # arguments:
+
+                def lossfunc(Qw, nodeID, dummyl, dummygrid):
+                    return float(loss_function(Qw, nodeID))
+
+                self._lossfunc = lossfunc
+
+            elif num_params == 3:
+                # check the func takes (float, int) and turns it into a new
+                # single value:
+                if not isinstance(loss_function(1., 0, 0), float):
+                    raise TypeError(
+                        "The loss_function should take (float, int, int), "
+                        "and return a float."
+                    )
+
+                def lossfunc(Qw, nodeID, linkID, dummygrid):
+                    return float(loss_function(Qw, nodeID, linkID))
+
+                self._lossfunc = lossfunc
+
+            elif num_params == 4:
+                # this time, the test is too hard to implement cleanly so just
+                self._lossfunc = loss_function
+            else:
+                raise ValueError(
+                    "The loss_function must have only a single argument, "
+                    "which should be the discharge at a node; a pair of "
+                    "arguments, which should be the discharge at a node and "
+                    "the node ID; or three arguments, which should be the "
+                    "discharge at a node, the node ID, and the link along "
+                    "which that discharge will flow."
+                )
+        else:
+            # make a dummy
+            def lossfunc(Qw, dummyn, dummyl, dummygrid):
+                return float(Qw)
+
+            self._lossfunc = lossfunc
+
+        # add the new loss discharge field if necessary:
+        if "surface_water__discharge_loss" not in grid.at_link:
+            self.grid.add_zeros(
+                "node", "surface_water__discharge_loss", dtype=float, noclobber=False
+            )
+
```

Comment:
@SiccarPoint a thought about this field -- that we should discuss on the next call. 

This field JUST records the impact of the loss function (direction of information: from function to component to field). But, lets pretend that there is someone who doesn't want/need to use a loss function, but wants to specify a spatially variable field that is "surface_water__discharge_loss" (or "surface_water__discharge_loss_proportion"?)

We should at least consider supporting this. I've made notes in the teleconference agenda.

Reply:
I like this a lot, but it should certainly be a separate PR if we want to support it. I'll leave this open to remind me to make one once we've resolved the naming as above. I think it would be slightly clearer to have this as an absolute loss rather than a proportional one, though I'm not wedded to that. Basically, I think it will save a round of calculation if you're taking this as an output and doing something else with it. Clearly if an absolute number we'd have to threshold it at the amount of discharge currently at the node if used as an input.

Reply:
We're suggesting on the call that the input field should be specified as a FLUX i.e. [L/T] to make it obviously distinct from the output field - also many inputs might provide this.

Will make a PR.